extendr-api/src/scalar/rint.rs

```diff
@@ -0,0 +1,207 @@
+use super::super::IsNA;
+use std::ops::{Add, Div, Mul, Neg, Not, Sub};
+
+/// Rint is a wrapper for i32 in the context of an R's integer vector.
+///
+/// Rint can have a value between i32::MIN+1 and i32::MAX
+///
+/// The value i32::MIN is used as "NA".
+///
+/// Rint has the same footprint as an i32 value allowing us to use it in zero copy slices.
+#[derive(PartialEq, Eq)]
+pub struct Rint(pub i32);
+
+impl Rint {
+    /// Construct a NA Rint.
+    pub fn na() -> Self {
+        Rint(i32::MIN)
+    }
+}
+
+impl Clone for Rint {
+    fn clone(&self) -> Self {
+        Self(self.0)
+    }
+}
+
+impl Copy for Rint {}
+
+impl IsNA for Rint {
+    /// Return true is the is a NA value.
+    fn is_na(&self) -> bool {
+        self.0 == std::i32::MIN
+    }
+}
+
+impl From<i32> for Rint {
+    /// Construct a Rint from an integer.
+    /// i32::MIN gives an NA.
+    fn from(v: i32) -> Self {
+        Self(v)
+    }
+}
+
+impl From<Option<i32>> for Rint {
+    /// Construct an Rint from an optional integer.
+    /// None or Some(i32::MIN) gives an NA.
+    fn from(v: Option<i32>) -> Self {
+        if let Some(v) = v {
+            v.into()
+        } else {
+            Rint::na()
+        }
+    }
+}
+
+impl From<Rint> for Option<i32> {
+    /// Convert an Rint to an optional integer.
+    /// NA gives None.
+    fn from(v: Rint) -> Self {
+        if v.is_na() {
+            None
+        } else {
+            Some(v.0)
+        }
+    }
+}
+
+impl std::fmt::Debug for Rint {
+    /// Debug format an Rint.
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let z: Option<i32> = (*self).into();
+        if let Some(val) = z {
+            write!(f, "{}", val)
+        } else {
+            write!(f, "na")
+        }
+    }
+}
+
+macro_rules! gen_binop {
```

Comment:
Could you add just one or two sentences here commenting what this does? E.g., something like: "Programmatically derive binary operators"? Also, if this hardcodes the `Rint` type maybe call it `gen_binop_Rint`, since we'll need something similar for `Rfloat`?